control-plane/connect-inject/controllers/pod/pod.go

```diff
@@ -0,0 +1,361 @@
+// has been configured with and will only delete tokens for the provided podName.
+// func (r *Controller) deleteACLTokensForWorkload(apiClient *api.Client, svc *api.AgentService, k8sNS, podName string) error {
+
+// TODO: add support for explicit upstreams
+//func (r *Controller) writeUpstreams(pod corev1.Pod) error
+
+// shouldIgnore ignores namespaces where we don't mesh-inject.
+func shouldIgnore(namespace string, denySet, allowSet mapset.Set) bool {
+	// Ignores system namespaces.
+	if namespace == metav1.NamespaceSystem || namespace == metav1.NamespacePublic || namespace == "local-path-storage" {
+		return true
+	}
+
+	// Ignores deny list.
+	if denySet.Contains(namespace) {
+		return true
+	}
+
+	// Ignores if not in allow list or allow list is not *.
+	if !allowSet.Contains("*") && !allowSet.Contains(namespace) {
+		return true
+	}
+
+	return false
+}
+
+// consulNamespace returns the Consul destination namespace for a provided Kubernetes namespace
+// depending on Consul Namespaces being enabled and the value of namespace mirroring.
+func (r *Controller) getConsulNamespace(kubeNamespace string) string {
+	// TODO: use namespaces.ConsulNamespace in the future where this field isn't mandatory
+	if !r.EnableConsulNamespaces {
+		return constants.DefaultConsulNS
+	}
+
+	// Mirroring takes precedence.
+	if r.EnableNSMirroring {
+		return fmt.Sprintf("%s%s", r.NSMirroringPrefix, kubeNamespace)
+	}
+
+	return r.ConsulDestinationNamespace
+}
+
+func (r *Controller) getPartition() string {
+	if !r.EnableConsulPartitions || r.Partition == "" {
+		return constants.DefaultConsulPartition
+	}
+	return r.Partition
+}
+
+func getWorkloadPorts(pod corev1.Pod) ([]string, map[string]*pbcatalog.WorkloadPort) {
+	ports := make([]string, 0)
+	workloadPorts := map[string]*pbcatalog.WorkloadPort{}
+
+	for _, container := range pod.Spec.Containers {
+		for _, port := range container.Ports {
+			name := port.Name
+			if name == "" {
+				name = strconv.Itoa(int(port.ContainerPort))
+			}
+
+			// TODO: error check reserved "mesh" keyword and 20000
+
+			if port.Protocol != corev1.ProtocolTCP {
+				// TODO: also throw an error here
+				continue
+			}
+
+			ports = append(ports, name)
+			workloadPorts[name] = &pbcatalog.WorkloadPort{
+				Port: uint32(port.ContainerPort),
+
+				// We leave the protocol unspecified so that it can be inherited from the Service appProtocol
+				Protocol: pbcatalog.Protocol_PROTOCOL_UNSPECIFIED,
+			}
+		}
+	}
+
+	ports = append(ports, "mesh")
+	workloadPorts["mesh"] = &pbcatalog.WorkloadPort{
+		Port:     constants.ProxyDefaultInboundPort,
+		Protocol: pbcatalog.Protocol_PROTOCOL_MESH,
+	}
+
+	return ports, workloadPorts
+}
+
+func parseLocality(node corev1.Node) *pbcatalog.Locality {
+	region := node.Labels[corev1.LabelTopologyRegion]
+	zone := node.Labels[corev1.LabelTopologyZone]
+
+	if region == "" {
+		return nil
+	}
+
+	return &pbcatalog.Locality{
+		Region: region,
+		Zone:   zone,
+	}
+}
```

Comment:
This differs slightly from the one in endpoints since it returns a `pbcatalog.Locality`, and I'm not sure any other controllers need it, so I'll keep it as-is for now.

Reply:
Ah, yeah that's fair. SGTM!